In [1]:
from scipy import stats
import numpy as np

https://www.khanacademy.org/math/statistics-probability/analysis-of-variance-anova-library

*Key Questions*
1. How much variation comes from within the groups compared to between groups?
2. Are the true population means between groups the same?


*Setup*
 - 3 groups (a, b, c) each containing 3 observations (m = 3, n = 3)

In [2]:
a = np.array([3,2,1])
b = np.array([5,3,4])
c = np.array([5,6,7])
m = 3
n = 3

In [3]:
grand_mean = np.sum([a,b,c])/(m*n)
sst = np.sum([(x-grand_mean)**2 for x in [a,b,c]])
df = (m*n)-1
grand_mean, sst, df

(4.0, 30.0, 8)

In [4]:
ssw = np.sum([(ss - np.mean(s))**2 for s in [a,b,c] for ss in s])
dfw = m*(n-1) #obs per group times groups - 1
ssw, dfw

(6.0, 6)

In [5]:
ssb = np.sum([len(s)*((np.mean(s) - grand_mean)**2) for s in [a,b,c]])
dfb = m - 1 #obs per group - 1
ssb, dfb

(24.0, 2)

H0: All the population means are the same mu1=m2=m3
H1: Means are not the same

Assume H0:
F-statistic = ( SSB / DFB ) / (SSW / DFW )

 - If the numerator is larger, then that means there is a greater difference BETWEEN, so less chance that our null hypothesis is correct.

 - If the numerator is smaller, then that means there is a greater difference WITHIN, so better chance that our null hypothesis is correct.

In [6]:
alpha = 0.1
f_stat = (ssb/dfb)/(ssw/dfw)
critical_f = stats.f.ppf(1-alpha, dfb, dfw)
f_stat, dfb, dfw, critical_f

(12.0, 2, 6, 3.4633040700956514)

Check F Distribution Tables: http://www.socr.ucla.edu/Applets.dir/F_Table.html

For alpha = 0.10
df numerator = 2
df denominator = 6
Critical F = 3.46

Given 12 is much higher than our critical f-statistic, we REJECT the null hypotheis that our population means are equal.
So we can conclude that there is a difference between the groups.